Feat: add update_window to workspace_group for create/update

pbr0ck3r opened the pull request.

pbr0ck3r commented:
Getting the following error when applying and assume it is because the API might not be returning `update_window` or allowing it to be passed in via the `management.go` library. 

```When applying changes to singlestoredb_workspace_group.this, provider "provider[\"registry.terraform.io/hashicorp/singlestoredb\"]" produced an unexpected new value: .update_window: was cty.ObjectVal(map[string]cty.Value{"day":cty.NumberIntVal(6), "hour":cty.NumberIntVal(17)}), but now null.```

noprysk-ua commented on .tool-versions:
```diff
@@ -0,0 +1 @@
+golang 1.20.5
```
What is the `.tool-versions` file? I'm not familiar with it and not sure what would be the value from including it. I would expect `go.mod` to be enough.

pbr0ck3r replied:
`.tool-versions` file is used by package managers such as `asdf` to allow you to run different versions of packages per project. I can remove it.

noprysk-ua reviewed (commented):
Hi Phil,

Thank you for creating the MR that adds the update window functionality to the workspace group resource!

The current state of the MR is a solid foundation.

1. Our team is working on setting update window details on the creation of a workspace group and this MR is blocked on the feature because, otherwise, the behavior of the provider is not consistent. E.g., an operator indicates Sunday yet gets a random day and overrides the random day only on the second `terraform apply` run. I'll make sure to notify you once the functionality is ready.

2. The MR only implements updating the state, yet does not implement reading the state from remote. The function `toWorkspaceGroupResourceModel` is the relevant converter that should be changed. It's essential to provide proper reading as it will enable the proper resolution of deltas.

3. By the convention of the current repository, almost all the functionality is covered by unit and integration tests. Make sure to extend the MR with both unit and integration tests. Note that it would be enough to extend tests that already exist.

Best,
Nestor

pbr0ck3r commented on internal/provider/workspacegroups/resource.go:
```diff
@@ -36,13 +36,14 @@ type workspaceGroupResource struct {
+	ExpiresAt      types.String                 `tfsdk:"expires_at"`
+	RegionID       types.String                 `tfsdk:"region_id"`
+	AdminPassword  types.String                 `tfsdk:"admin_password"`
+	UpdateWindow   *updateWindowDataSourceModel `tfsdk:"update_window"`
```
I will make the changes you requested, thanks for the feedback!

pbr0ck3r replied:
Updated the `ImportState` and also added `updateWindowResourceModel`. Will adjust testing and look forward to your team updating the operator.

pbr0ck3r commented:
> The updateWindow input for a workspace group creation is now live. Moreover, the latest version of the Golang client for the Management API is pulled to this repo.

Has the operator code been deployed yet? Began adjusting tests and still receive a random day/hour.

noprysk-ua commented:
> > The updateWindow input for a workspace group creation is now live. Moreover, the latest version of the Golang client for the Management API is pulled to this repo.
> 
> Has the operator code been deployed yet? Began adjusting tests and still receive a random day/hour.

Yes. The current version of the PR does not pass day and hour as input. This is a possible reason for the confusion.

I've just verified that all works fine with a `curl` request like this.

```bash
curl -X POST "https://api.singlestore.com/v1/workspaceGroups" -H "accept: application/json" -H "Authorization: Bearer $API_KEY" -H "Content-Type: application/json" -d "{\"adminPassword\":\"qw5erASDF19\",\"firewallRanges\":[],\"name\":\"demo-workspace-group\",\"regionID\":\"7e7ffd27-20f7-44b6-87e6-e72828a81ac7\", \"updateWindow\":{\"day\":1,\"hour\":2}}"
{"workspaceGroupID":"936580ff-215f-4e43-b0bd-2ec1a5561ddc"}
curl -X GET "https://api.singlestore.com/v1/workspaceGroups" -H "accept: application/json" -H "Authorization: Bearer $API_KEY"
[{"name":"demo-workspace-group","state":"PENDING","workspaceGroupID":"936580ff-215f-4e43-b0bd-2ec1a5561ddc","createdAt":"2023-07-07T14:53:57.007404Z","regionID":"7e7ffd27-20f7-44b6-87e6-e72828a81ac7","updateWindow":{"hour":2,"day":1}}]
```

Note that `hour` is 2, just as indicated in input and the `day` is 1, which is consistent with the input as well.

pbr0ck3r commented:
@noprysk-ua , went ahead and updated the tests. I could also test the helper functions `toManagementUpdateWindow` and `toUpdateWindowResourceModel` I added in a follow up PR similar to the `util` tests.
